language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/packages/DocumentComponentsVisitor.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://wso2.com) All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.extensions.ballerina.packages;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import io.ballerina.compiler.syntax.tree.ClassDefinitionNode;
+import io.ballerina.compiler.syntax.tree.ConstantDeclarationNode;
+import io.ballerina.compiler.syntax.tree.EnumDeclarationNode;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.ListenerDeclarationNode;
+import io.ballerina.compiler.syntax.tree.ModuleVariableDeclarationNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.TypeDefinitionNode;
+import io.ballerina.tools.text.LineRange;
+
+import java.util.stream.Collectors;
+
+/**
+ * The components Visitor class to get the list of module level components.
+ */
+public class DocumentComponentsVisitor extends NodeVisitor {
+    private final JsonObject module;
+
+    private JsonObject currentService = null;
+    private JsonObject currentClass = null;
+    private boolean isInsideClass;
+    private boolean isInsideService;
+
+    DocumentComponentsVisitor(JsonObject module) {
+        this.module = module;
+    }
+
+    public void getDocumentComponents(Node node) {
+        visitSyntaxNode(node);
+    }
```

Comment:
Here we are maintaining a state in between the construct visits which is not a good approach with a visitor like this. Specially dynamically setting variables like `isInsideClass` and `isInsideService` might not be a good approach since we might end up with filling invalid information.

Instead of the visitor, let's use a node transformer.